discovery/openstack/instance.go

```diff
@@ -198,7 +205,7 @@ func (i *InstanceDiscovery) refresh(ctx context.Context) ([]*targetgroup.Group,
 					}
 					lbls[openstackLabelAddressPool] = model.LabelValue(pool)
 					lbls[openstackLabelPrivateIP] = model.LabelValue(addr)
-					if val, ok := floatingIPList[floatingIPKey{id: s.ID, fixed: addr}]; ok {
+					if val, ok := floatingIPList[floatingIPKey{tenantID: s.TenantID, fixed: addr}]; ok {
```

Comment:
I'm afraid we no longer identify the concerned instance/server as precisely as before.
Two instances with the same fixed private IP will both satisfy this condition even though only one of them has a floating IP.

Reply:
@machine424 sorry for this time without news 🙏 

I understand your point, and you're right since we can have multiple networks with the same CIDR block in the same tenant and this will create a concurrency issue.

We need to refactor this and rely on the `/ports` API since the [Floating IP List method.](https://github.com/gophercloud/gophercloud/blob/v1/openstack/networking/v2/extensions/layer3/floatingips/results.go#L17-L59) from Gophercloud won't give the `port_details`, which usually contains the `device_id`, where the Floating IP is attached to correlate data within the instance ID.

Follow my structural proposal before taking it into code:
- Build a map with the `ID`, `FloatingIP`, `FixedIP`, `TenantID`, and `PortID` from the `/floatingips` API.
- Build a map with the `ID`, `TenantID`, and `DeviceID` from the `/ports` API.
  - On this map, only keep entries in which the `ID` exists on the `PortID` Floating IPs map.

With this we can do on the instance addresses loop the proper and unique match to a floating IP based on the `FixedIP` (`addr`) and the `DeviceID`  from the Ports map since this value will be the `ID` from the instance.

Let me know what you think.

\cc @simonpasquier

Reply:
Hey @paulojmdias, thanks for giving me a heads-up on this PR.

I think I did more or less what you propose in mine: https://github.com/prometheus/prometheus/pull/15558
I am using the port ID alone because I expect the UUID to be unique; however that implies refactoring the second part where the function relied on the server `addresses`; I propose to replace the whole thing with a call to the `ports` endpoint. PTAL!

Note that if you're still working on this PR, I am happy to let you deal with the conversion to Neutron FIPs and I will rebase to just bump the Gophercloud version.

Reply:
Hey @pierreprinetti,

Yeah, I think we will need the `ports` endpoint to ensure we can avoid any possible concurrency issue.

If you agree, I will work on that on this PR, and you can rebase afterward to bump the Gophercloud version.

Either way, I also have opened #15539, which if the team agrees, I think makes sense to be dependent on the bump of the Gophercloud version. After that, I can rebase to use it on that PR.

Reply:
Sounds excellent to me, thanks! I'll wait for this #14367 to merge then.

Reply:
@pierreprinetti @machine424 The PR is already updated with the code needed to rely on `/ports` API for the right Floating IPs association to the instances.

I also added 2 new instances to the mock tests using the same private IP in different tenants to test proper discovery.

If you feel anything is missing, please let me know.